server/src/main/java/org/elasticsearch/script/AbstractFieldScript.java

```diff
@@ -79,13 +81,16 @@ public AbstractFieldScript(
         LeafReaderContext ctx,
         OnScriptError onScriptError
     ) {
-        super(new DocValuesDocReader(searchLookup, ctx));
+        this(fieldName, params, new DocValuesDocReader(searchLookup, ctx), onScriptError);
+    }
+
+    private AbstractFieldScript(String fieldName, Map<String, Object> params, DocReader docReader, OnScriptError onScriptError) {
+        super(docReader);
         this.fieldName = fieldName;
-        Map<String, Object> docAsMap = docAsMap();
-        this.sourceLookup = (SourceLookup) docAsMap.get("_source");
+        this.source = docReader.source();
         params = new HashMap<>(params);
-        params.put("_source", sourceLookup);
-        params.put("_fields", docAsMap.get("_fields"));
+        params.put("_source", this.source);
```

Comment:
Will scripts be ok with this change? I think they may expect something that implements `Map`.

Reply:
The DynamicMap implementations handle converting it so that we only reify the source as a Map if a script directly calls it.

Reply:
Neat